### PR TITLE
Layers

### PR DESCRIPTION
**Pull request by tomlepaine**
@BigRedT @kevjshih 

I added some layer helper functions. This is basically how I make layers in tf. Any suggestions are welcome, from re-naming functions, and changing variable names, to adding comments, to better tests.

The main point of the layers is to take the hassle out of creating weight params, and initializing them properly. It also makes the code that uses them look a bit cleaner. Something like

```
x = tf.placeholder(tf.float32, [batch, dim])
h1 = full(x, 10, 'h1')
h2 = full(h1, 10, 'h2')
h3 = full(h2, 10, 'h3')
y_hat = full(h3, 10, 'y_hat',
    gain=1.0,
    func=identity) # This is a linear layer with properly initialized weights.
```

This should also keep the **tensorboard** graphs cleaner.

---

I also wrote a simple test that I guess does three things in one:
- checks that the layers don't throw any exceptions.
- implicitly checks the shape out the output.
- insures that zero inputs give zero outputs, which should be the case for the default initialization values.

Best,
Tom

### CODE REVIEW

**Comment by kevjshih** [comment]
Looks good. Should we create a layers subdirectory in tftools for this?

**Comment by tomlepaine** [comment]
I think we probably should eventually, but how about we wait until need too? Like after we have 10 layers or something.

**Comment by kevjshih** [comment]
Sure, I think we should just push it in asap then. Go ahead and merge if you think it's ready, though it might be a good idea to run it through the code format checker you suggested in the email thread first.

File: aiutils/tftools/layers.py [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+
+
+def identity(input):
+    return input
+
+
+def full(input, out_dim, name, gain=np.sqrt(2), func=tf.nn.relu):
+    """ Fully connected layer helper.
+    
+    Creates weights and bias parameters with good initial values. Then applies the matmul op and func.
```
Comment (BigRedT) [inline aiutils/tftools/layers.py]: Might as well mention xavier glorot initialization.

File: aiutils/tftools/layers.py [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+      func (function): Function used to calculate neural activations.
+      
+    Returns:
+      output (tensor): The neural activations for this layer.
+        Will have shape `[batch, out_dim]`.
+    """
+    in_dim = input.get_shape().as_list()[-1]
+    stddev = 1.0 * gain / np.sqrt(in_dim)
+    with tf.variable_scope(name):
+        w_init = tf.random_normal_initializer(stddev)
```
Comment (BigRedT) [inline aiutils/tftools/layers.py]: Probably change to truncated normal?

File: aiutils/tftools/layers.py [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+      func (function): Function used to calculate neural activations.
+      
+    Returns:
+      output (tensor): The neural activations for this layer.
+        Will have shape `[batch, in_weight, in_width, out_dim]`.
+    """
+    in_dim = input.get_shape().as_list()[-1]
+    stddev = 1.0 * gain / np.sqrt(filter_size * filter_size * in_dim)
+    with tf.variable_scope(name):
+        w_init = tf.random_normal_initializer(stddev)
```
Comment (BigRedT) [inline aiutils/tftools/layers.py]: truncated normal here as well?

File: aiutils/tftools/layers.py [hunk 4]
```diff
@@ -0,0 +1,83 @@
+import numpy as np
+import tensorflow as tf
+
+
+def identity(input):
```
Comment (BigRedT) [inline aiutils/tftools/layers.py]: what is the purpose of identity?

Comment (kevjshih) [inline aiutils/tftools/layers.py]: I agree it's probably not that useful unless it were a means of initialization (Identity initialized RNNs for example)? I'd probably take it out in its current form.

**Comment by BigRedT** [comment]
@tomlepaine: I added some comments to you code.

**Pull request merged**